File: archive/scripts/create_bucket.py
import boto3
from botocore.exceptions import ClientError
import os


def create_bucket(s3_url, s3_access_key, s3_secret_key, bucket_name):
    """
    Create a bucket on the MinIO server using the AWS S3 API.
    """
    # Initialize S3 client
    s3_client = boto3.client(
        "s3",
        endpoint_url=s3_url,
        aws_access_key_id=s3_access_key,
        aws_secret_access_key=s3_secret_key,
    )

    try:
        # Check if the bucket already exists
        response = s3_client.list_buckets()
        for bucket in response.get("Buckets", []):
            if bucket["Name"] == bucket_name:
                print(f"Bucket '{bucket_name}' already exists.")
                return

        # Create the bucket
        s3_client.create_bucket(Bucket=bucket_name)
        print(f"Bucket '{bucket_name}' created successfully!")
    except ClientError as e:
        print(f"Error creating bucket: {e}")


if __name__ == "__main__":
    # MinIO server configuration
    MINIO_URL = os.getenv("MINIO_URL", "http://localhost:9000")
    ACCESS_KEY = os.getenv("MINIO_ROOT_USER", "ROOT_USER")
    SECRET_KEY = os.getenv("MINIO_ROOT_PASSWORD", "TOOR_PASSWORD")
    BUCKET_NAME = os.getenv("MINIO_BUCKET_NAME", "papers")
    create_bucket(
        s3_url=MINIO_URL,
        s3_access_key=ACCESS_KEY,
        s3_secret_key=SECRET_KEY,
        bucket_name=BUCKET_NAME,
    )

File: scripts/createBucket.ts
// filepath: scripts/createBucket.ts
import { S3Client, ListBucketsCommand, CreateBucketCommand } from "@aws-sdk/client-s3"

// MinIO/S3 server configuration from environment variables
const MINIO_URL = process.env.MINIO_URL || "http://localhost:9000"
const ACCESS_KEY = process.env.MINIO_ROOT_USER || "ROOT_USER"
const SECRET_KEY = process.env.MINIO_ROOT_PASSWORD || "TOOR_PASSWORD"
const BUCKET_NAME = process.env.MINIO_BUCKET_NAME || "papers"
const REGION = process.env.MINIO_REGION || "us-east-1"

// Configure the AWS SDK v3 S3Client
const s3 = new S3Client({
    endpoint: MINIO_URL,
    region: REGION,
    credentials: {
        accessKeyId: ACCESS_KEY,
        secretAccessKey: SECRET_KEY,
    },
    forcePathStyle: true, // Required for MinIO
})

async function createBucketIfNotExists(bucketName: string) {
    try {
        // Check if the bucket already exists
        const listResult = await s3.send(new ListBucketsCommand({}))
        const bucketExists = (listResult.Buckets || []).some(
            (bucket) => bucket.Name === bucketName
        )

        if (bucketExists) {
            console.log(`Bucket '${bucketName}' already exists.`)
        } else {
            await s3.send(new CreateBucketCommand({ Bucket: bucketName }))
            console.log(`Bucket '${bucketName}' created successfully!`)
        }
    } catch (err: any) {
        if (err && typeof err === "object" && "message" in err) {
            console.error(`Error interacting with S3: ${err.message}`)
        } else {
            console.error("Error interacting with S3:", err)
        }
    }
}

// Run the function
createBucketIfNotExists(BUCKET_NAME)